src/import.js

```diff
@@ -339,88 +310,95 @@ Sk.importModuleInternal_ = function (name, dumpJS, modname, suppliedPyBody, canS
+        var codeAndPath, co, googClosure, external;
+        var searchFileName = name;
+        var result;
+
+        topLevelModuleToReturn = topLevelModuleToReturn_;
+
+        // If we're inside a package, look search using its __path__
+        if (modNameSplit.length > 1) {
+            parentModule = Sk.sysmodules.mp$subscript(absolutePackagePrefix + parentModName);
+            searchFileName = modNameSplit[modNameSplit.length-1];
+            searchPath = parentModule.tp$getattr("__path__");
         }
 
-        // check first for an externally loaded library
-        external = Sk.loadExternalLibrary(name);
-        if (external) {
-            co = external;
-            if (Sk.externalLibraries) {
-                filename = Sk.externalLibraries[name].path; // get path from config
-            } else {
-                filename = "unknown";
-            }
-            // ToDo: check if this is a dotted name or import from ...
+        // otherwise:
+        // - create module object
+        // - add module object to sys.modules
+        // - compile source to (function(){...});
+        // - run module and set the module locals returned to the module __dict__
+        module = new Sk.builtin.module();
+
+        if (suppliedPyBody) {
+            filename = name + ".py";
+            co = Sk.compile(suppliedPyBody, filename, "exec", canSuspend);
         } else {
-            // Try loading as a builtin (i.e. already in JS) module, then try .py files
-            codeAndPath = Sk.importSearchPathForName(name, ".js", true, canSuspend, currentDir);
-
-            co = (function compileReadCode(codeAndPath) {
-                if (codeAndPath instanceof Sk.misceval.Suspension) {
-                    return new Sk.misceval.Suspension(compileReadCode, codeAndPath);
-                } else if (!codeAndPath) {
-                    goog.asserts.assert(!isPy, "Sk.importReadFileFromPath did not throw when loading Python file failed");
-                    isPy = true;
-                    return compileReadCode(Sk.importSearchPathForName(name, ".py", false, canSuspend, currentDir));
+            // If an onBeforeImport method is supplied, call it and if
+            // the result is false or a string, prevent the import.
+            // This allows for a user to conditionally prevent the usage
+            // of certain libraries.
+            if (Sk.onBeforeImport && typeof Sk.onBeforeImport === "function") {
+                result = Sk.onBeforeImport(name);
+                if (result === false) {
+                    throw new Sk.builtin.ImportError("Importing " + name + " is not allowed");
+                } else if (typeof result === "string") {
+                    throw new Sk.builtin.ImportError(result);
+                }
+            }
+
+            // check first for an externally loaded library
+            external = Sk.loadExternalLibrary(name);
```

Comment:
should this not be handled by `Sk.read`? Not in the scope of this PR just wondering if you agree.

Reply:
Yes, I would like that. We should have a chat about external code, because one of these days I'm implementing AOT compilation for Anvil...